net/gateway/gateway.go

```diff
@@ -0,0 +1,157 @@
+package gateway
+
+import (
+	"fmt"
+	"go.sia.tech/core/net/mux"
+	"go.sia.tech/core/net/rpc"
+	"go.sia.tech/core/types"
+	"io"
+	"net"
+	"sync"
+)
+
+type GatewayID [8]byte
+
+type Session struct {
+	uniqueID   GatewayID
+	netAddress NetAddress
+
+	err    error
+	closed bool
+	mu     sync.Mutex
+
+	mux *mux.Mux
+}
+
+func (s *Session) IsClosed() bool {
+	return s.closed
+}
+
+func (s *Session) Close() error {
+	if s.IsClosed() {
+		return nil
+	}
+
+	s.mu.Lock()
+	s.closed = true
+	s.mu.Unlock()
+
+	return s.mux.Close()
+}
+
+// PrematureCloseErr returns the error that resulted in the Session being closed
+// prematurely.
+func (s *Session) PrematureCloseErr() error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.err
+}
+
+func (s *Session) setErr(err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err != nil && s.err == nil {
+		if ne, ok := err.(net.Error); !ok || !ne.Temporary() {
+			s.err = err
+			s.mux.Close()
+		}
+	}
+}
+
+func (s *Session) writeMessage(obj rpc.ProtocolObject) error {
+	if err := s.PrematureCloseErr(); err != nil {
+		return err
+	}
+
+	stream, err := s.mux.DialStream()
```

Comment:
we want a 1-to-1 mapping between mux streams and RPCs; the flow for an RPC is:
1. create a new mux stream
2. send rpc ID
3. send rpc message body
4. read rpc response
5. close mux stream

So I think the right approach is to move the `DialStream` into `WriteMessage`, and have `writeMessage` operate on the stream.

Reply:
It's an additional string handshake:
https://gitlab.com/NebulousLabs/siamux/-/blob/master/siamux.go#L301
https://github.com/SiaFoundation/siad/blob/master/modules/host/network.go#L250

https://github.com/lukechampine/us/pull/178